File: typescript/packages/core/src/session-router.ts
import type { AgentRunRequest, TerminalReason } from './types/domain.js';
import type { AgentLoop, EventBus, RunManager, SessionLane, SessionRouter, StreamEvent } from './types/contracts.js';

export class DefaultSessionRouter implements SessionRouter {
  constructor(
    private readonly lane: SessionLane,
    private readonly loop: AgentLoop,
    private readonly runs: RunManager,
    private readonly bus: EventBus,
  ) {}

  async *route(request: AgentRunRequest): AsyncGenerator<StreamEvent, TerminalReason, unknown> {
    const run = this.runs.create(request.sessionId);
    this.runs.update(run.runId, { status: 'running' });

    const queue: StreamEvent[] = [];
    let finished = false;
    let wake: (() => void) | undefined;
    const state: { terminal: TerminalReason } = { terminal: { kind: 'completed' } };

    const signal = (): void => {
      wake?.();
      wake = undefined;
    };

    void this.lane.enqueue(request.sessionId, async () => {
      try {
        const generator = this.loop.run(request);
        let result = await generator.next();
        while (!result.done) {
          this.bus.publish(result.value);
          queue.push(result.value);
          signal();
          result = await generator.next();
        }
        state.terminal = result.value ?? { kind: 'completed' };
      } catch (error) {
        state.terminal = {
          kind: 'error',
          message: error instanceof Error ? error.message : String(error),
        };
      } finally {
        finished = true;
        signal();
      }
    });

    while (true) {
      while (queue.length > 0) {
        yield queue.shift()!;
      }
      if (finished) {
        break;
      }
      await new Promise<void>((resolve) => {
        wake = resolve;
        queueMicrotask(() => {
          if (queue.length > 0 || finished) {
            resolve();
          }
        });
      });
    }

    const terminal = state.terminal;
    const status =
      terminal.kind === 'error'
        ? 'failed'
        : terminal.kind === 'cancelled'
          ? 'cancelled'
          : 'completed';

    this.runs.update(run.runId, {
      status,
      endedAt: new Date().toISOString(),
      error: terminal.kind === 'error' ? terminal.message : undefined,
    });

    return terminal;
  }
}

File: python/src/agent_framework/core/session_router.py
from __future__ import annotations

import asyncio
from datetime import datetime, timezone
from typing import AsyncIterator

from agent_framework.core.contracts import AgentLoop, EventBus, LifecycleEvent, RunManager, SessionLane, StreamEvent
from agent_framework.core.domain import AgentRunRequest


class DefaultSessionRouter:
    def __init__(self, lane: SessionLane, loop: AgentLoop, runs: RunManager, bus: EventBus) -> None:
        self._lane = lane
        self._loop = loop
        self._runs = runs
        self._bus = bus

    async def route(self, request: AgentRunRequest) -> AsyncIterator[StreamEvent]:
        run = self._runs.create(request.session_id)
        self._runs.update(run.run_id, {"status": "running"})

        queue: asyncio.Queue[StreamEvent | None] = asyncio.Queue()

        async def producer() -> None:
            try:
                async for event in self._loop.run(request):
                    self._bus.publish(event)
                    await queue.put(event)
            except Exception as exc:  # noqa: BLE001
                await queue.put(LifecycleEvent(phase="error", run_id=run.run_id, error=str(exc)))
            finally:
                await queue.put(None)

        async def run_in_lane() -> None:
            await producer()

        lane_task = asyncio.create_task(self._lane.enqueue(request.session_id, run_in_lane))

        while True:
            event = await queue.get()
            if event is None:
                break
            yield event

        await lane_task

        self._runs.update(run.run_id, {
            "status": "completed",
            "ended_at": datetime.now(timezone.utc).isoformat(),
        })
